jbmc/src/java_bytecode/generic_parameter_specialization_map.h

```diff
@@ -0,0 +1,108 @@
+/// Author: Diffblue Ltd.
+
+#ifndef CPROVER_JAVA_BYTECODE_GENERIC_PARAMETER_SPECIALIZATION_MAP_H
+#define CPROVER_JAVA_BYTECODE_GENERIC_PARAMETER_SPECIALIZATION_MAP_H
+
+#include <stack>
+#include <vector>
+
+#include <util/string_utils.h>
+
+#include "expr2java.h"
+#include "java_types.h"
+
+class generic_parameter_specialization_mapt
+{
+private:
+  struct container_paramt
+  {
+    std::size_t container_index;
+    std::size_t param_index;
+  };
+  std::unordered_map<irep_idt, container_paramt> param_to_container;
+  std::vector<std::stack<std::vector<reference_typet>>>
+    container_to_specializations;
+
+public:
+  std::size_t insert(
+    const std::vector<java_generic_parametert> &parameters,
+    std::vector<reference_typet> types);
+
+  void pop(std::size_t container_id);
+
+  optionalt<reference_typet> pop(const irep_idt &parameter_name);
+
+  struct printert
+  {
+    const generic_parameter_specialization_mapt &map;
+    const namespacet &ns;
+
+    printert(
+      const generic_parameter_specialization_mapt &map,
+      const namespacet &ns)
+      : map(map), ns(ns)
+    {
+    }
+  };
+
+  template <typename ostream>
+  friend ostream &operator<<(ostream &stm, const printert &map);
+};
+
+template <typename ostream>
+ostream &operator<<(
+  ostream &stm,
```

Comment:
:question: Why does `ostream` need to be a template parameter instead of being just `std::ostream`?

Reply:
It could be restricted to `std::ostream` but it doesn't need to be. I've added comment to that effect.